langwatch/src/tasks/backfillScenarioAnalytics.ts

```diff
@@ -0,0 +1,190 @@
+import { prisma } from "~/server/db";
+import { esClient } from "~/server/elasticsearch";
+import type { AggregationsCalendarInterval } from "@elastic/elasticsearch/lib/api/types";
+import { ANALYTICS_KEYS } from "~/types";
+import { ScenarioEventType } from "~/app/api/scenario-events/[[...route]]/enums";
+import { createScenarioAnalyticsQueriesForAllEventTypes } from "~/server/scenario-analytics";
+
+interface DateHistogramBucket {
+  key_as_string: string;
+  doc_count: number;
+}
+
+export default async function execute() {
+  console.log("Starting backfillScenarioAnalytics...");
+
+  // Get all projects
+  const projects = await prisma.project.findMany({
+    select: {
+      id: true,
+      createdAt: true,
+    },
+  });
+
+  const client = await esClient({ test: true });
+  const today = new Date();
+  today.setHours(0, 0, 0, 0);
+
+  // For each project, backfill from creation date to today
+  for (const project of projects) {
+    try {
+      console.log(`\nProcessing project ${project.id}`);
+      const startDate = new Date(project.createdAt);
+      startDate.setHours(0, 0, 0, 0);
+
+      // Create multi-search body for all event types with date histogram aggregation
+      const msearchBody = createScenarioAnalyticsQueriesForAllEventTypes(
+        project.id,
+        startDate.getTime(),
+        today.getTime(),
+        true, // includeDateHistogram
```

Comment:
because of cases like this, I prefer to always used named params (ie an params object):

```
someFn(params: {
   projectId: string,
   startDateInMs: number,
   nowInMs: number,
   ...etc
})
```

this prevents bugs, IMO, and makes it easier to read. It also makes it easier to extend and better handles when you want something to be optional/undefined.